framework/include/utils/MortarUtils.h

```diff
@@ -0,0 +1,251 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "Assembly.h"
+#include "FEProblemBase.h"
+#include "SubProblem.h"
+
+#include "libmesh/quadrature.h"
+#include "libmesh/elem.h"
+#include "libmesh/point.h"
+
+namespace Moose
+{
+namespace Mortar
+{
+void projectQPoints3d(const Elem * msm_elem,
+                      const Elem * primal_elem,
+                      unsigned int sub_elem_index,
+                      const QBase & qrule_msm,
+                      std::vector<Point> & q_pts);
+
+template <typename Iterators, typename Consumers, typename ActionFunctor>
+void
+loopOverMortarSegments(const Iterators & secondary_elems_to_mortar_segments,
+                       Assembly & assembly,
+                       SubProblem & subproblem,
+                       FEProblemBase & fe_problem,
+                       const AutomaticMortarGeneration & amg,
+                       const bool displaced,
+                       const Consumers & consumers,
+                       const ActionFunctor act)
```

Comment:
What's the optimization resulting from leaving this implementation in the header file?

Reply:
Probably no optimization/inlining, but template function definitions generally need to be in the header unless you do explicit template instantiations in the .C file